Note that I get a test timeout when running tests on Nightly/Aurora, although they work on Beta/Release, and they behave the same on the master branch (so it isn't an issue specific to this branch):

```
Traceback (most recent call last):
  File "/Users/myk/Mozilla/r2d2b2g/addon-sdk/bin/cfx", line 33, in <module>
    cuddlefish.run()
  File "/Users/myk/Mozilla/r2d2b2g/addon-sdk/python-lib/cuddlefish/__init__.py", line 928, in run
    bundle_sdk=options.bundle_sdk)
  File "/Users/myk/Mozilla/r2d2b2g/addon-sdk/python-lib/cuddlefish/runner.py", line 743, in run_app
    OUTPUT_TIMEOUT, test_name, parseable)
Exception: Test output exceeded timeout (60s).
make: *** [test] Error 1
```

--- a/addon/lib/simulator.js
+++ b/addon/lib/simulator.js
@@ -845,10 +845,13 @@ let simulator = module.exports = {
 
     // NOTE: add errors/warnings for name and icons manifest attributes
     //       and updates name and icon attributes on the registered app object
-    simulator._validateNameIcons(app.validation.errors, app.validation.warnings,
-                                 app.manifest, app);
-    // NOTE: add errors/warnings for WebAPIs not supported by the simulator
-    simulator._validateWebAPIs(app.validation.errors, app.validation.warnings,
+    Validator.validateNameIcons(app.validation.errors, app.validation.warnings,
+                                app.manifest, app);
+    Validator.validatePermissions(app.validation.errors, app.validation.warnings,
+                                  app.manifest);
+    Validator.validateType(app.validation.errors, app.validation.warnings,
+                           app.manifest, app);
+    Validator.validateManifest(app.validation.errors, app.validation.warnings,
                                app.manifest);
 
     // Appcache checks
@@ -860,101 +863,13 @@ let simulator = module.exports = {
       app.validation.warnings.push("Packaged apps don't support appcache");
     }
 
-    if (["generated", "hosted"].indexOf(app.type) !== -1 &&
-        ["certified", "privileged"].indexOf(app.manifest.type) !== -1) {
-      app.validation.errors.push("Hosted App can't be type '" + app.manifest.type + "'.");
-      if (typeof next === "function") {
-        // NOTE: blocking error
-        next(Error("Invalid Manifest."), app);
-        return;
-      }
-    } else {
-      app.validation.running = true;
-      simulator.run(function (error) {
-        // on error running b2g-desktop, reports error and exits
-        if (error) {
-          if (typeof next === "function") {
-            app.validation.errors.push("Unable to complete manifest validation: " +
-                                       error);
-            next(Error("Unable to complete manifest validation."), app);
-          }
-          return;
-        }
-
-        simulator.remoteSimulator.validateManifest(app.manifest, function (reply) {
-          app.validation.running = false;
-          console.debug("VALIDATE REPLY: ", JSON.stringify(reply, null, 2));
-          if (reply.error) {
-            app.validation.errors.push("Unable to complete manifest validation: " +
-                                       reply.error);
-            next(Error("Unable to complete manifest validation."), app);
-            return;
-          }
-          if (!reply.success && reply.errors && reply.errors.length > 0) {
-            // concatenate validation errors as warnings (non-blocking errors)
-            app.validation.warnings = app.validation.warnings.
-              concat(reply.errors);
-          }
-          // check if there's any validation error
-          if (typeof next === "function") {
-            if (app.validation.errors.length === 0) {
-              next(null, app);
-            } else {
-              next(Error("Invalid Manifest."), app);
-            }
-          }
-        });
-      });
-    }
-  },
-
-  _validateNameIcons: function(errors, warnings, manifest, app) {
-      if (!manifest.name) {
-        errors.push("Missing mandatory 'name' in Manifest.");
-      }
-
-      if (!manifest.icons || Object.keys(manifest.icons).length == 0) {
-        warnings.push("Missing 'icons' in Manifest.");
+    // check if there's any validation error
+    if (typeof next === "function") {
+      if (app.validation.errors.length === 0) {
+        next(null, app);
       } else {
-        // update registered app icon
-        let size = Object.keys(manifest.icons).sort(function(a, b) b - a)[0] || null;
-        if (size) {
-          app.icon = manifest.icons[size];
-        }
-
-        // NOTE: add warnins if 128x128 icon is missing
-        if (! manifest.icons["128"]) {
-          warnings.push("app submission to the Marketplace needs at least an 128 icon");
-        }
+        next(Error("Invalid Manifest."), app);
       }
-
-      // update name visible in the dashboard
-      app.name = manifest.name;
-  },
-
-  _validateWebAPIs: function(errors, warnings, manifest) {
-    // certified app are not fully supported on the simulator
-    if (manifest.type === "certified") {
-      warnings.push("'certified' apps are not fully supported on the Simulator");
-    }
-
-    if (!manifest.permissions) {
-      return warnings;
-    }
-
-    let permissions = Object.keys(manifest.permissions);
-    let formatMessage = function (apiName) {
-      return "WebAPI '"+ apiName + "' is not currently supported on the Simulator";
-    };
-
-    // WebSMS is not currently supported on the simulator
-    if (permissions.indexOf("sms") > -1) {
-      warnings.push(formatMessage("WebSMS"));
-    }
-
-    // WebTelephony is not currently supported on the simulator
-    if (permissions.indexOf("telephony") > -1) {
-      warnings.push(formatMessage("WebTelephony"));
     }
   },
 

--- a/addon/lib/validator.js
+++ b/addon/lib/validator.js
@@ -5,15 +5,16 @@
 
 const { Ci, Cu, Cr } = require("chrome");
 
-const AppsUtils = Cu.import("resource://gre/modules/AppsUtils.jsm");
+const { AppsUtils, ManifestHelper } = Cu.import("resource://gre/modules/AppsUtils.jsm");
 const { Services } = Cu.import("resource://gre/modules/Services.jsm");
+const { PermissionsTable, expandPermissions } = Cu.import("resource://gre/modules/PermissionsTable.jsm");
 const { defer } = require('sdk/core/promise');
 
 exports.validateAppCache = function(errors, warnings, rawManifest, origin) {
   let deferred = defer();
 
   // Use ManifestHelper to normalize and retrieve the absolute appcache URI
-  let manifest = new AppsUtils.ManifestHelper(rawManifest, origin);
+  let manifest = new ManifestHelper(rawManifest, origin);
   if (!manifest.appcache_path) {
     deferred.resolve();
     return deferred.promise;
@@ -101,3 +102,107 @@ ChannelListener.prototype = {
     throw Cr.NS_ERROR_NO_INTERFACE;
   }
 };
+
+
+exports.validateNameIcons = function(errors, warnings, manifest, app) {
+  if (!manifest.name) {
+    errors.push("Missing mandatory 'name' in Manifest.");
+  }
+  // update name visible in the dashboard
+  app.name = manifest.name;
+
+  if (!manifest.icons || Object.keys(manifest.icons).length == 0) {
+    warnings.push("Missing 'icons' in Manifest.");
+  } else {
+    // update registered app icon
+    let size = Object.keys(manifest.icons).sort(function(a, b) b - a)[0] || null;
+    if (size) {
+      app.icon = manifest.icons[size];
+    }
+
+    // NOTE: add warnings if 128x128 icon is missing
+    if (!manifest.icons["128"]) {
+      warnings.push("app submission to the Marketplace needs at least an 128 icon");
+    }
+  }
+}
+
+exports.validateManifest = function (errors, warnings, manifest) {
+  let valid = AppsUtils.checkManifest(manifest, {});
+
+  if (!valid) {
+    errors.push("This app can't be installed on a production device "+
+                "(AppsUtils.checkManifest return false).");
+  }
+}
+
+exports.validateType = function (errors, warnings, manifest, app) {
+  let appType = manifest.type || "web";
+  if (["web", "privileged", "certified"].indexOf(appType) === -1) {
+    errors.push("Unknown app type: '" + appType + "'.");
+  } else if (["generated", "hosted"].indexOf(app.type) !== -1 &&
+             ["certified", "privileged"].indexOf(manifest.type) !== -1) {
+    errors.push("Hosted App can't be type '" + manifest.type + "'.");
+  }
+
+  // certified app are not fully supported on the simulator
+  if (manifest.type === "certified") {
+    warnings.push("'certified' apps are not fully supported on the Simulator.");
+  }
+}
+
+exports.validatePermissions = function(errors, warnings, manifest) {
+  if (!manifest.permissions) {
+    return;
+  }
+
+  let permissionsNames = Object.keys(manifest.permissions);
+
+  let formatMessage = function (apiName) {
+    return "WebAPI '"+ apiName + "' is not currently supported on the Simulator";
+  };
+
+  // WebSMS is not currently supported on the simulator
+  if (permissionsNames.indexOf("sms") > -1) {
+    warnings.push(formatMessage("WebSMS"));
+  }
+
+  // WebTelephony is not currently supported on the simulator
+  if (permissionsNames.indexOf("telephony") > -1) {
+    warnings.push(formatMessage("WebTelephony"));
+  }
+
+  let appType = manifest.type || "web";
+  let appStatus;
+  // NOTE: If it isn't certified or privileged, it's appStatus "app"
+  // https://hg.mozilla.org/releases/mozilla-b2g18/file/d9278721eea1/dom/apps/src/PermissionsTable.jsm#l413
+  if (["privileged", "certified"].indexOf(appType) === -1) {
+    appStatus = "app";
+  } else {
+    appStatus = appType;
+  }
+
+  permissionsNames.forEach(function(name) {
+    let permission = PermissionsTable[name];
+
+    if (permission) {
+      let permissionAction = permission[appStatus];
+      if (!permissionAction) {
+        errors.push("Ignored permission '" + name + "' (invalid app type '" + appType + "').");
+      } else if (permissionAction === Ci.nsIPermissionManager.DENY_ACTION) {
+        errors.push("Denied permission '" + name + "' for app type '" + appType + "'.");
+      } else {
+        let access = manifest.permissions[name].access;
+        try {
+          if (access && expandPermissions(name, access).length === 0) {
+            errors.push("Invalid access '" + access + "' in permission '" + name + "'.");
+          }
+        } catch(e) {
+          errors.push("Invalid access '" + access + "' in permission '" + name + "'.");
+        }
+      }
+    } else {
+      errors.push("Unknown permission '" + name + "'.");
+    }
+  });
+}

--- a/addon/test/test-validator.js
+++ b/addon/test/test-validator.js
@@ -31,7 +31,7 @@ exports["test validate empty appcache"] = function(assert, done) {
   });
 }
 
-exports["test relative path"] = function(assert, done) {
+exports["test relative appcache path"] = function(assert, done) {
   let errors = [], warnings = [];
   let manifest = {appcache_path: "appcache.manifest"};
   let promise = validator.validateAppCache(errors, warnings, manifest, origin);
@@ -42,7 +42,7 @@ exports["test relative path"] = function(assert, done) {
   })
 }
 
-exports["test absolute path"] = function(assert, done) {
+exports["test absolute appcache path"] = function(assert, done) {
   let errors = [], warnings = [];
   let manifest = {appcache_path: "/appcache.manifest"};
   let promise = validator.validateAppCache(errors, warnings, manifest, origin);
@@ -53,7 +53,7 @@ exports["test absolute path"] = function(assert, done) {
   });
 }
 
-exports["test absolute URI"] = function(assert, done) {
+exports["test absolute apccache URI"] = function(assert, done) {
   let errors = [], warnings = [];
   let manifest = {appcache_path: origin + "/appcache.manifest"};
   let promise = validator.validateAppCache(errors, warnings, manifest, origin);
@@ -64,7 +64,7 @@ exports["test absolute URI"] = function(assert, done) {
   });
 }
 
-exports["test redirected"] = function(assert, done) {
+exports["test redirected appcache"] = function(assert, done) {
   server.registerPathHandler("/redirected.manifest", function handle(request, response) {
     response.setStatusLine(request.httpVersion, 301, "Moved Permanently");
     response.setHeader("Location", origin + "/appcache.manifest", false);
@@ -84,7 +84,7 @@ exports["test redirected"] = function(assert, done) {
   });
 }
 
-exports["test nonexistent"] = function(assert, done) {
+exports["test nonexistent appcache"] = function(assert, done) {
   server.registerPathHandler("/nonexistent.manifest", function handle(request, response) {
     response.setStatusLine(request.httpVersion, 404, "Not Found");
   });
@@ -100,6 +100,176 @@ exports["test nonexistent"] = function(assert, done) {
     assert.equal(warnings.length, 0, "no warning for absolute URI");
     done();
   });
+}
+
+exports["test empty manifest"] = function(assert) {
+  let errors = [], warnings = [];
+  let manifest = {}, app = {};
+  validator.validateNameIcons(errors, warnings, manifest, app);
+  assert.equal(errors.length, 1, "missing name in manifest is an error");
+  assert.equal(errors[0], "Missing mandatory 'name' in Manifest.");
+  assert.equal(warnings.length, 1, "missing icons in manifest causes a warning");
+  assert.equal(warnings[0], "Missing 'icons' in Manifest.");
+}
+
+exports["test biggest icon selection"] = function(assert) {
+  let errors = [], warnings = [];
+  let manifest = {
+    name: "foo",
+    icons: {"16": "16.png", "128": "128.png", "32": "32.png"}
+  };
+  let app = {};
+  validator.validateNameIcons(errors, warnings, manifest, app);
+  assert.equal(errors.length, 0);
+  assert.equal(warnings.length, 0);
+  assert.equal(app.name, "foo", "app.name has been set");
+  assert.equal(app.icon, "128.png", "selected the biggest icon");
+}
+
+exports["test marketplace icon size condition"] = function(assert) {
+  let errors = [], warnings = [];
+  let manifest = {
+    name: "foo",
+    icons: {"16": "16.png"}
+  };
+  let app = {};
+  validator.validateNameIcons(errors, warnings, manifest, app);
+  assert.equal(errors.length, 0);
+  assert.equal(warnings.length, 1);
+  assert.equal(warnings[0], "app submission to the Marketplace needs at least an 128 icon");
+}
+
+exports["test privileged hosted"] = function(assert) {
+  let errors = [], warnings = [];
+  let manifest = {
+    type: "privileged"
+  };
+  let app = {
+    type: "hosted"
+  };
+  validator.validateType(errors, warnings, manifest, app);
+  assert.equal(errors.length, 1);
+  assert.equal(errors[0], "Hosted App can't be type 'privileged'.");
+}
+
+exports["test unknown type"] = function(assert) {
+  let errors = [], warnings = [];
+  let manifest = {
+    type: "privilegeddd"
+  };
+  let app = {};
+  validator.validateType(errors, warnings, manifest, app);
+  assert.equal(errors.length, 1);
+  assert.equal(errors[0], "Unknown app type: 'privilegeddd'.");
+}
+
+exports["test certified warning"] = function(assert) {
+  let errors = [], warnings = [];
+  let manifest = {
+    type: "certified"
+  };
+  let app = {
+    type: "packaged"
+  };
+  validator.validateType(errors, warnings, manifest, app);
+  assert.equal(warnings.length, 1);
+  assert.equal(warnings[0], "'certified' apps are not fully supported on the Simulator.");
+}
+
+exports["test simulator unsupported permission"] = function(assert) {
+  let errors = [], warnings = [];
+  let manifest = {
+    permissions: {sms : {}}
+  };
+  let app = {};
+  validator.validatePermissions(errors, warnings, manifest, app);
+  assert.equal(warnings.length, 1);
+  assert.equal(warnings[0], "WebAPI 'WebSMS' is not currently supported on the Simulator");
+}
+
+exports["test certified only permission - accepted"] = function(assert) {
+  let errors = [], warnings = [];
+  let manifest = {
+    permissions: {mobileconnection : {}},
+    type: "certified"
+  };
+  let app = {};
+  validator.validatePermissions(errors, warnings, manifest, app);
+  assert.equal(errors.length, 0);
+  assert.equal(warnings.length, 0);
+}
+
+exports["test certified only permission - denied"] = function(assert) {
+  let errors = [], warnings = [];
+  let manifest = {
+    permissions: {mobileconnection : {}}
+  };
+  let app = {};
+  validator.validatePermissions(errors, warnings, manifest, app);
+  assert.equal(errors.length, 1);
+  assert.equal(errors[0], "Denied permission 'mobileconnection' for app type 'web'.");
+}
+
+exports["test permission access"] = function(assert) {
+  let errors = [], warnings = [];
+  let manifest = {
+    permissions: {"device-storage:videos" : {access: "readonly"}},
+    type: "privileged"
+  };
+  let app = {};
+  validator.validatePermissions(errors, warnings, manifest, app);
+  assert.equal(errors.length, 0);
+  assert.equal(warnings.length, 0);
+}
+
+exports["test permission wrong access"] = function(assert) {
+  // Typo in access
+  let errors = [], warnings = [];
+  let manifest = {
+    permissions: {"device-storage:videos" : {access: "readonlyyy"}},
+    type: "privileged"
+  };
+  let app = {};
+  validator.validatePermissions(errors, warnings, manifest, app);
+  assert.equal(errors.length, 1);
+  assert.equal(errors[0], "Invalid access 'readonlyyy' in permission 'device-storage:videos'.");
+}
+
+exports["test permission access on permission without access"] = function(assert) {
+  let errors = [], warnings = [];
+  let manifest = {
+    permissions: {"alarms" : {access: "readonly"}}
+  };
+  let app = {};
+  validator.validatePermissions(errors, warnings, manifest, app);
+  assert.equal(errors.length, 1);
+  assert.equal(errors[0], "Invalid access 'readonly' in permission 'alarms'.");
+}
+
+exports["test permission denied access"] = function(assert) {
+  // Use an access field that doesn't match the targeted permission accesses
+  let errors = [], warnings = [];
+  let manifest = {
+    permissions: {"device-storage:apps" : {access: "createonly"}},
+    type: "certified"
+  };
+  let app = {};
+  validator.validatePermissions(errors, warnings, manifest, app);
+  assert.equal(errors.length, 1);
+  assert.equal(errors[0], "Invalid access 'createonly' in permission 'device-storage:apps'.");
+}
+
+exports["test checkManifest - non-numeric size"] = function(assert) {
+  let errors = [], warnings = [];
+  let manifest = {
+    name: "foo",
+    size: "non-numeric-size"
+  };
+  let app = {};
+  validator.validateManifest(errors, warnings, manifest, app);
+  assert.equal(errors.length, 1);
+  assert.equal(errors[0], "This app can't be installed on a production device " +
+                          "(AppsUtils.checkManifest return false).");
 }
 
 exports["test z teardown"] = function (assert, done) {

--- a/prosthesis/content/dbg-simulator-actors.js
+++ b/prosthesis/content/dbg-simulator-actors.js
@@ -221,83 +221,6 @@ SimulatorActor.prototype = {
     };
   },
 
-  onValidateManifest: function(aRequest) {
-    this.debug("simulator actor received 'validateManifest' command: " + JSON.stringify(aRequest));
-    let manifest = aRequest.manifest;
-    let appType = manifest.type || "web";
-
-    let errors = [];
-
-    if (["web", "privileged", "certified"].indexOf(appType) === -1) {
-      errors.push("Unknown app type: '" + appType + "'.");
-    }
-
-    let utils = {};
-    Cu.import("resource://gre/modules/AppsUtils.jsm", utils);
-    let valid = utils.AppsUtils.checkManifest(manifest, {});
-
-    if (!valid) {
-      errors.push("This app can't be installed on a production device "+
-                  "(AppsUtils.checkManifest return false).");
-    }
-
-    if (manifest.permissions) {
-      this._validateManifestPermissions(appType, manifest.permissions, errors);
-    }
-
-    if (errors.length > 0) {
-      return {
-        success: false,
-        errors: errors
-      };
-    }
-
-    return {
-      success: true
-    };
-  },
-
-  _validateManifestPermissions: function(appType, permissions, errors) {
-    let utils = {};
-    Cu.import("resource://gre/modules/PermissionsTable.jsm", utils);
-
-    let permissionsNames = Object.keys(permissions);
-
-    let appStatus;
-    // NOTE: If it isn't certified or privileged, it's appStatus "app"
-    // https://hg.mozilla.org/releases/mozilla-b2g18/file/d9278721eea1/dom/apps/src/PermissionsTable.jsm#l413
-    if (["privileged", "certified"].indexOf(appType) === -1) {
-      appStatus = "app";
-    } else {
-      appStatus = appType;
-    }
-
-    permissionsNames.forEach(function(pname) {
-      let permission = utils.PermissionsTable[pname];
-
-      if (permission) {
-        let permissionAction = permission[appStatus];
-        if (!permissionAction) {
-          errors.push("Ignored permission '" + pname + "' (invalid type '" + appType + "').");
-        } else if (permissionAction === Ci.nsIPermissionManager.DENY_ACTION) {
-          errors.push("Denied permission '" + pname + "' for type '" + appType + "'.");
-        } else {
-          let access = permissions[pname].access;
-          try {
-            if (access && utils.expandPermissions(pname, access).length === 0) {
-              errors.push("Invalid access '" + access + "' in permission '" + pname + "'.");
-            }
-          } catch(e) {
-            this.debug(["EXCEPTION:", e, e.fileName, e.lineNumber].join(' '));
-            errors.push("Invalid access '" + paccess + "' in permission '" + pname + "'.");
-          }
-        }
-      } else {
-        errors.push("Unknown permission '" + pname + "'.");
-      }
-    });
-  },
-
   onUninstallApp: function(aRequest) {
     this.debug("simulator actor received 'uninstallApp' command: " + aRequest.appId);
     let window = this.simulatorWindow;
@@ -384,7 +307,6 @@ SimulatorActor.prototype.requestTypes = {
   "getBuildID": SimulatorActor.prototype.onGetBuildID,
   "runApp": SimulatorActor.prototype.onRunApp,
   "uninstallApp": SimulatorActor.prototype.onUninstallApp,
-  "validateManifest": SimulatorActor.prototype.onValidateManifest,
   "showNotification": SimulatorActor.prototype.onShowNotification,
   "geolocationUpdate": SimulatorActor.prototype.onGeolocationUpdate,
 };

